packages/commerce-sdk-react/src/hooks/types.ts

```diff
@@ -38,35 +50,91 @@ export interface ApiClients {
+export type Argument<T extends (arg: any) => unknown> = NonNullable<Parameters<T>[0]>
 
 /**
  * The data type returned by a commerce-sdk-isomorphic method when the raw response
  * flag is not set.
  */
-export type DataType<T extends (arg: any) => Promise<unknown>> = T extends (
-    arg: any
-) => Promise<Response | infer R>
-    ? R
-    : never
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type DataType<T> = T extends ApiMethod<any, Response | infer R> ? R : never
 
 /**
- * Modified version of React Query's Mutation Function. Added a second argument
- * API clients.
+ * Merged headers and parameters from client config and options, mimicking the behavior
+ * of commerce-sdk-isomorphic.
  */
-export type IMutationFunction<TData = unknown, TVariables = unknown> = (
-    variables: TVariables,
-    apiClients: ApiClients
-) => Promise<TData>
+export type MergedOptions<Client extends ApiClient, Options extends ApiOptions> = RequireKeys<
+    ApiOptions<
+        NonNullable<Client['clientConfig']['parameters'] & Options['parameters']>,
+        NonNullable<Client['clientConfig']['headers'] & Options['headers']>,
+        // `body` may not exist on `Options`, in which case it is `unknown` here. Due to the type
+        // constraint in `ApiOptions`, that is not a valid value. We must replace it with `never`
+        // to indicate that the result type does not have a `body`.
+        unknown extends Options['body'] ? never : Options['body']
+    >,
+    'parameters' | 'headers'
+>
 
-/**
- * Modified version of React Query's Query Function. Added a second argument
- * API clients.
- */
-export type IQueryFunction<TData = unknown> = (
-    context: QueryFunctionContext<QueryKey>,
-    apiClients: ApiClients
-) => Promise<TData>
+// --- CACHE HELPERS --- //
+
+export type ApiQueryKey =
+    // | readonly string[] // TODO: Is this needed?
+    readonly [...path: string[], parameters: Record<string, unknown>]
+
+export type CacheUpdateUpdate<T> = {
+    queryKey: ApiQueryKey
+    updater: Updater<T | undefined, T | undefined>
+}
+
+export type CacheUpdateInvalidate = (query: Query) => boolean
+
+export type CacheUpdateRemove = (query: Query) => boolean
+
+export type CacheUpdate = {
+    update?: CacheUpdateUpdate<unknown>[]
+    invalidate?: CacheUpdateInvalidate[]
+    remove?: CacheUpdateRemove[]
+}
+
+export type CacheUpdateGetter<Options, Data> = (
+    customerId: string | null,
+    options: Options,
+    response: Data
+) => CacheUpdate
+
+export type CacheUpdateMatrix<Client extends ApiClient> = {
+    // It feels like we should be able to do <infer Arg, infer Data>, but that
+    // results in some methods being `never`, so we just use Argument<> later
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [Method in keyof Client]?: Client[Method] extends ApiMethod<any, Response | infer Data>
+        ? CacheUpdateGetter<MergedOptions<Client, Argument<Client[Method]>>, Data>
+        : never
+}
```

Comment:
Does this look scary? Probably! But basically what it says is "for each method on the client, the cache update matrix has an optional method that accepts a `CacheUpdateGetter` with that method's argument and return type".